Add test update self

iranika opened the pull request:
Continuation of # 167
Added test update self in tester.nim .

genotrance commented on tests/tester.nim:
```diff
@@ -171,3 +172,31 @@ when defined(linux):
+
+    (output, exitCode) = exec(["update","self","--debug"], liveOutput=true)
+    check exitCode == QuitSuccess
+    check inLines(output.processOutput, "Info: Already up to date at version")
```
This is checking if the current shipping version of choosenim, that was just updated to, can update. It isn't testing the currently built code. Not sure if there's a point in testing this and if it can break the CI in the future.

iranika replied:
I should be testing for already updated, before build lower version info.
I don't know risk too, it can break the CI in the future. I want some good ideas.

iranika replied:
Is it better to add a force option to update self?🤔

genotrance commented on tests/tester.nim:
```diff
@@ -171,3 +172,31 @@ when defined(linux):
+    check exitCode == QuitSuccess
+    check inLines(output.processOutput, "Info: Already up to date at version")
+    block cleanup:
+      removeFile(rootDir / "bin" / "choosenim_0.4.0") #remove lower version exefile.
```
Where does the `_0.4.0` suffix come from? `nimble build` doesn't append this to the name.

iranika replied:
This is created by updateSelf().
https://github.com/dom96/choosenim/blob/0e7b2c16d4912f508161393d5b1f974eec7d4027/src/choosenim.nim#L92
Old execute files are renamed.

dom96 commented on tests/tester.nim:
```diff
@@ -171,3 +172,34 @@ when defined(linux):
+
+test "can update self":
+  beginTest()
+  var (output, exitCode) = exec(["update","self","--debug"], liveOutput=true)
```
Nit: spaces after comma

iranika commented:
@dom96 @genotrance 
I added the '--force' option, And rewrote test. I pass Travis test in my branch, But This Pull Reqest failed.
[My branch test result](
https://travis-ci.org/iranika/choosenim/builds/647167264?utm_medium=notification&utm_source=github_status)
[Pull Request test result](https://travis-ci.org/dom96/choosenim/builds/647158838?utm_medium=notification&utm_source=github_status)
I'm not sure now, but I think it is due to `.travis.yml`.  #170
I will examine it later 🤕

genotrance commented:
@iranika - please hold off on testing this PR, am working on fixing the build script. Also, join us on #nim if possible so that we can chat.

iranika commented:
@genotrance OK, I hold off on testing this PR, wait # 168 to complete.
I will sleep. Thanks!

dom96 commented on tests/tester.nim:
```diff
@@ -171,3 +172,21 @@ when defined(linux):
+  block :
+    let (output, exitCode) = exec(["update", "self", "--debug"], liveOutput=true)
+    check exitCode == QuitSuccess
+    check inLines(output.processOutput, "Info: Already up to date at version")
```
This depends on the external state, which means it will be flaky as soon as we try to run the test suite on an older version of choosenim (which may happen). I think this can be removed.

iranika replied:
OK, I'll remove it.

dom96 commented on tests/tester.nim:
```diff
@@ -171,3 +172,21 @@ when defined(linux):
+      # moveFile don't overwritten on windows. So, delete it.
+      when defined(windows): removeFile(exePath)
+      moveFile(rootDir / "bin" / "choosenim_" & chooseNimVersion.addFileExt(ExeExt),
+               exePath)
```
Why is this needed?

iranika replied:
Because, Test Changes should be undone.
[Line 184](https://github.com/dom96/choosenim/pull/169/files#diff-0e23be51b85eef33ca502065a5cb52e0R184)```let (output, exitCode) = exec(["update", "self", "--debug", "--force"], liveOutput=true)```
This will replace the Exec File  in the bin with the latest `release` version. For subsequent tests, the release version is used, not current build. It's unintended.
And, the artifacts in the bin directory are deployed by TravisCI. In this state, the current build will not be deployed.
So we need back to the state before the test.

iranika replied:
@dom96 I understood why bin dir is used, in this case.
`updateSelf()` doesn't use options `--choosenimDir` and ` --nimbleDir`. It's used getAppFilename().
https://github.com/dom96/choosenim/blob/25f86483552c99c3c70f5271d8a5123f238b9d26/src/choosenim.nim#L91-L102
So, release binary is downloaded to bin dir , in this case.
My mind has changed. Now, I think I should copy `bin/choosenim` to `tests/choosenimDir/` dir, and test using `tests/choosenimDir/choosenim`. In this case, bin dir  has no changed. It's clean. :)
Do you have any thoughts?

dom96 replied:
Yeah, I think that sounds better. Although I have a vague memory of how this fits together :)

iranika replied:
@dom96 I done this fix. No trash in the bin dir : )
evidence:
![image](https://user-images.githubusercontent.com/7370486/74404929-2d813400-4e6f-11ea-8115-705debe02247.png)
